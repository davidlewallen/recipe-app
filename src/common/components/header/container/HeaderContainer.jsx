// @flow
import React from 'react';

import { type RouterHistory } from 'react-router-dom';
import { type Recipes, type Recipe } from '../../../types';

import { Account, Utils } from '../../../utils/api';

import Header from '../components';
import SubmitRecipeContainer from '../../submit-recipe/container/SubmitRecipeContainer';
import AcceptedWebsites from '../../acceptedWebsites/components';

import '../assets/styles/index.css';

type Props = {
  history: RouterHistory,
  updateRecipes: (recipes: Recipes | Recipe) => void,
  recipes: Recipes,
  isAuth: boolean,
  updateAuth: () => void,
}

type State = {
  showModal: boolean,
  showAcceptedModal: boolean,
  acceptedWebsites: Array<String>,
}

class HeaderContainer extends React.Component<Props, State> {
  state = {
    showModal: false,
    showAcceptedModal: false,
    acceptedWebsites: [],
  };

  componentWillMount = async () => {
    const { data: acceptedWebsites } = await Utils.getAcceptedWebsites();
    this.setState({ acceptedWebsites });
  }

  logout = () => {
    Account.logout();
    this.props.updateAuth(false);
    this.props.history.replace('/');
  }

  handleModalOpen = () => {
    this.setState({ showModal: true });
  }

  handleModalClose = () => {
    this.setState({ showModal: false });
  }

  handleAcceptedModal = () => {
    this.setState(prevState => (
      { showAcceptedModal: !prevState.showAcceptedModal }
    ));
  }

  render = () => (
    <div>
      <SubmitRecipeContainer
        show={this.state.showModal}
        handleModalClose={this.handleModalClose}
        updateRecipes={this.props.updateRecipes}
        recipes={this.props.recipes}
      />
      <AcceptedWebsites
        show={this.state.showAcceptedModal}
        handleAcceptedModal={this.handleAcceptedModal}
        acceptedWebsites={this.state.acceptedWebsites}
      />
      <Header
        logout={this.logout}
        handleModalOpen={this.handleModalOpen}
        isAuth={this.props.isAuth}
        handleAcceptedModal={this.handleAcceptedModal}
      />
    </div>
  );
}

export default HeaderContainer;
