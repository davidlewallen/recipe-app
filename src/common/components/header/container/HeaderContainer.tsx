import * as React from 'react';

import { IPropTypes, IState } from './types';

import { Account, Utils } from '../../../utils/api';

import Header from '../components';
import SubmitRecipeContainer from '../../submit-recipe/container/SubmitRecipeContainer';
import AcceptedWebsites from '../../acceptedWebsites/components';

import '../assets/styles/index.css';

class HeaderContainer extends React.Component<IPropTypes, IState> {
  state = {
    showModal: false,
    showAcceptedModal: false,
    acceptedWebsites: [],
  };

  componentDidMount = async () => {
    const { data: acceptedWebsites } = await Utils.getAcceptedWebsites();

    this.setState({ acceptedWebsites });
  }

  logout = () => {
    const { updateAuth, history } = this.props;

    Account.logout();
    updateAuth(false);
    history.replace('/');
  }

  handleModalOpen = () => this.setState({ showModal: true });

  handleModalClose = () => this.setState({ showModal: false });

  handleAcceptedModal = () => (
    this.setState(prevState => ({ showAcceptedModal: !prevState.showAcceptedModal }))
  );

  render = () => {
    const {
      props: { updateRecipes, recipes, isAuth },
      state: { showModal, showAcceptedModal, acceptedWebsites },
    } = this;

    return (
      <React.Fragment>
        <SubmitRecipeContainer
          show={showModal}
          handleModalClose={this.handleModalClose}
          updateRecipes={updateRecipes}
          recipes={recipes}
        />
        <AcceptedWebsites
          show={showAcceptedModal}
          handleAcceptedModal={this.handleAcceptedModal}
          acceptedWebsites={acceptedWebsites}
        />
        <Header
          logout={this.logout}
          handleModalOpen={this.handleModalOpen}
          isAuth={isAuth}
          handleAcceptedModal={this.handleAcceptedModal}
        />
      </React.Fragment>
    );
  }
}

export default HeaderContainer;
