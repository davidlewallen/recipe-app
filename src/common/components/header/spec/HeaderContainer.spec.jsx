import React from 'react';
import { shallow } from 'enzyme';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import HeaderContainer from '../container/HeaderContainer';

const mock = new MockAdapter(axios);

describe('HeaderContainer', () => {
  const mockHistory = { history: { replace: jest.fn() } };

  const props = {
    ...mockHistory,
    updateRecipes: jest.fn(),
    recipes: [],
    isAuth: true,
    updateAuth: jest.fn(),
  };

  let wrapper = null;
  let instance = null;

  beforeEach(() => {
    jest.resetAllMocks();

    mock.reset();

    wrapper = shallow(<HeaderContainer {...props} />);
    instance = wrapper.instance();
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('logout', () => {
    it('should call logout', () => {
      mock.onGet('/api/account/logout').reply(200);
      const spy = jest.spyOn(instance, 'logout');
      instance.logout();
      expect(spy).toHaveBeenCalled();
    });

    it('should redirect to \'/\'', async () => {
      mock.onGet('/api/account/logout').reply(200);
      await instance.logout();
      expect(instance.props.history.replace).toHaveBeenCalledWith('/');
    });
  });
});
