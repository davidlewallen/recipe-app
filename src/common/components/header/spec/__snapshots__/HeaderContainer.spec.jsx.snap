// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderContainer should render 1`] = `
<Fragment>
  <SubmitRecipeContainer
    handleModalClose={[Function]}
    recipes={Array []}
    show={false}
    updateRecipes={[MockFunction]}
  />
  <AcceptedWebsites
    acceptedWebsites={Array []}
    handleAcceptedModal={[Function]}
    show={false}
  />
  <Header
    handleAcceptedModal={[Function]}
    handleModalOpen={[Function]}
    isAuth={true}
    logout={[Function]}
  />
</Fragment>
`;
